Highlighting in the doc that square root comes before adding epsilon

vincentqb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26735 Highlighting in the doc that square root comes before adding epsilon**

Differential Revision: [D17558505](https://our.internmc.facebook.com/intern/diff/D17558505)

vincentqb committed:
Highlighting in the documentation that square root comes before adding epsilon. https://github.com/pytorch/pytorch/issues/23796

vincentqb committed:
Update on "Highlighting in the doc that square root comes before adding epsilon"

zou3519 commented on torch/optim/rmsprop.py:
```diff
@@ -11,6 +11,9 @@ class RMSprop(Optimizer):
     With Recurrent Neural Networks <https://arxiv.org/pdf/1308.0850v5.pdf>`_.
 
+    Note that the implementation here takes the square root of the gradient average before
+    adding epsilon.
```
nits: It would be nice to have an equation like one of those that you showed me offline; equations are easier to read sometimes. Also, it might be good to document that this differs from the TF implementation specifically because I'm not sure if someone using RMSprop knows exactly where TF does the square root without investigation.

vincentqb replied:
Good points. I've added a note about tensorflow and the effective learning rate for RMSProp as `\alpha/(\sqrt{v} + \epsilon)`.

zou3519 reviewed (approved):
lgtm.

vincentqb committed:
Update on "Highlighting in the doc that square root comes before adding epsilon"

vincentqb committed:
Update on "Highlighting in the doc that square root comes before adding epsilon"


Differential Revision: [D17558505](https://our.internmc.facebook.com/intern/diff/D17558505)

zou3519 commented on torch/optim/rmsprop.py:
```diff
@@ -3,14 +3,20 @@
     With Recurrent Neural Networks <https://arxiv.org/pdf/1308.0850v5.pdf>`_.
 
+    The implementation here takes the square root of the gradient average before
+    adding epsilon (note that Tensor Flow interchange these two operations). The effective
```
TensorFlow is one word, I think. Also, I think this should be "note that TensorFlow interchanges these two operations"

vincentqb replied:
Oops! Thanks :)

vincentqb committed:
Update on "Highlighting in the doc that square root comes before adding epsilon"


Differential Revision: [D17558505](https://our.internmc.facebook.com/intern/diff/D17558505)

bonlime commented:
@vincentqb @zou3519 
Hi, 
May I ask is there any particular reason the current implementation of RMSprop adds epsilon after sqrt? 
From my experience training with default PyTorch RMSprop + FP16 (using apex) tends to produce `nan` during backprop. I'm sure @rwightman also experienced such issues. Training EfficientNets, for example, is impossible. 

By moving epsilon inside sqrt everything works like a charm. There are also mathematical proofs that adding eps after sqrt is not enough to prevent overflow. [Check this](https://arxiv.org/pdf/1911.05920.pdf) paper at p.6 bottom right. The topic of the paper is slightly different but their conclusion is also applicable here.

bonlime commented:
According to https://github.com/pytorch/pytorch/issues/23796 other people also had problems with RMSprop behavior in PyTorch. As far as I see you want the current implementation to be consistent with Adam / AdamW but they also tend to produce `nan`'s! 
It's a very tricky issue and it's hard to show that proposed implementation is superior to the current one by tests. But I know a bunch of people who successfully used adjusted versions of Adam / RMSprop to avoid nans
